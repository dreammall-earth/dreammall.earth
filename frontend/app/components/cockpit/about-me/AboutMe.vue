<template>
  <ClientOnly>
    <AboutMeView
      v-if="user && mode === 'view'"
      :username="user.username"
      :name="user.name"
      :availability="user.availability"
      :introduction="user.introduction"
      :details="user.details"
      :social="user.social"
      :user-image="avatar"
      :initials="initials"
      @edit-details="() => setMode('edit-details')"
      @edit-social="() => setMode('edit-social')"
      @update-availability="updateAvailability"
      @update-name="updateName"
      @update-introduction="updateIntroduction"
    />
    <EditUserDetails
      v-if="user && mode === 'edit-details'"
      :details="user.details"
      :loading="loading"
      @back="() => setMode('view')"
      @add-detail="addDetail"
      @remove-detail="removeDetail"
    />
    <EditSocialMedia
      v-if="user && mode === 'edit-social'"
      :socials="user.social"
      :loading="loading"
      @back="() => setMode('view')"
      @add-social="addSocial"
      @remove-social="removeSocial"
    />
  </ClientOnly>
</template>

<script lang="ts" setup>
import { storeToRefs } from 'pinia'
import { ref } from 'vue'

import ClientOnly from '#app/components/ClientOnly.vue'
import {
  useUserStore,
  UserAvailability,
  AddUserDetailInput,
  AddSocialMediaInput,
} from '#app/stores/userStore'
import globalErrorHandler from '#renderer/plugins/globalErrorHandler'

import AboutMeView from './AboutMeView.vue'
import EditSocialMedia from './EditSocialMedia.vue'
import EditUserDetails from './EditUserDetails.vue'

type Mode = 'view' | 'edit-details' | 'edit-social'
const mode = ref<Mode>('view')

const setMode = (newMode: Mode) => (mode.value = newMode)

const userStore = useUserStore()

const {
  getCurrentUser: user,
  getCurrentUserAvatar: avatar,
  getCurrentUserInitials: initials,
  loading,
} = storeToRefs(userStore)

const updateAvailability = async (newAvailability: UserAvailability) => {
  try {
    await userStore.updateUser({
      name: user.value!.name,
      availability: newAvailability,
    })
  } catch (error) {
    globalErrorHandler.error(`Could not update user: ${(error as Error).message}`, error)
  }
}

const updateName = async (newName: string) => {
  await userStore.updateUser({
    name: newName,
  })
}

const updateIntroduction = async (newIntroduction: string) => {
  await userStore.updateUser({
    name: user.value!.name,
    introduction: newIntroduction,
  })
}

const addDetail = async (detail: AddUserDetailInput) => {
  await userStore.addUserDetail(detail)
  detail.text = ''
}

const removeDetail = async (detailId: number) => {
  await userStore.removeUserDetail(detailId)
}

const addSocial = async (social: AddSocialMediaInput) => {
  await userStore.addSocialMedia(social)
  social.link = ''
}

const removeSocial = async (socialId: number) => {
  await userStore.removeSocialMedia(socialId)
}
</script>
