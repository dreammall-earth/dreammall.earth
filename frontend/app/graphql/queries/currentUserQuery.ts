import { gql } from 'graphql-tag'

export const currentUserQuery = gql`
  query {
    currentUser {
      id
      referenceId
      name
      username
      introduction
      availability
      details {
        id
        category
        text
      }
      social {
        id
        type
        link
      }
      table {
        id
        name
        public
        users {
          id
          role
          name
          username
        }
      }
    }
  }
`
