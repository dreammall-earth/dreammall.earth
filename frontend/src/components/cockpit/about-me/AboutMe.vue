<template>
  <CockpitCard>
    <template #header>
      <h2>{{ $t('cockpit.about-me.header') }}</h2>
    </template>
    <template #default>
      <!-- eslint-disable-next-line @intlify/vue-i18n/no-raw-text -->
      <div class="about-me-text">"{{ user.text }}"</div>
      <hr />
      <ul class="details">
        <li v-if="user.place"><v-icon icon="$place"></v-icon> {{ user.place }}</li>
        <li v-if="user.work"><v-icon icon="$working"></v-icon> {{ user.work }}</li>
        <li></li>
      </ul>
      <ul class="social">
        <li v-if="user.social.facebook">
          <a :href="user.social.facebook" target="_blank" rel="noopener noreferrer">
            <v-icon icon="$facebook"></v-icon>
          </a>
        </li>
        <li v-if="user.social.linkedin">
          <a :href="user.social.linkedin" target="_blank" rel="noopener noreferrer">
            <v-icon icon="$linkedIn"></v-icon>
          </a>
        </li>
        <li v-if="user.social.xing">
          <a :href="user.social.xing" target="_blank" rel="noopener noreferrer">
            <v-icon icon="$xing"></v-icon>
          </a>
        </li>
        <li v-if="user.social.x">
          <a :href="user.social.x" target="_blank" rel="noopener noreferrer">
            <v-icon icon="$x"></v-icon>
          </a>
        </li>
      </ul>
    </template>
  </CockpitCard>
</template>

<script lang="ts" setup>
import CockpitCard from '#components/cockpit/cockpitCard/CockpitCard.vue'

type User = {
  name: string
  text?: string
  place?: string
  work?: string
  social: {
    facebook?: string
    linkedin?: string
    xing?: string
    x?: string
  }
}

defineProps<{
  user: User
}>()
</script>

<style scoped lang="scss">
.about-me-text {
  overflow-wrap: break-word;
}

.details,
.social {
  padding: 0;
  list-style: none;
}

.social {
  display: flex;
}

hr {
  margin-block: 10px;
  border-color: white;
}
</style>
