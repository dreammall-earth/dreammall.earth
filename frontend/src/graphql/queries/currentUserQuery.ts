import { gql } from 'graphql-tag'

export const currentUserQuery = gql`
  query {
    currentUser {
      id
      name
      username
      room {
        id
        name
        public
        users {
          id
          role
          name
          username
        }
      }
    }
  }
`
