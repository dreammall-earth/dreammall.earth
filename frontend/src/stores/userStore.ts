import { useMutation, useQuery } from '@vue/apollo-composable'
import { acceptHMRUpdate, defineStore } from 'pinia'
import { computed, ref, watch } from 'vue'

import {
  addSocialMediaMutation,
  AddSocialMediaInput,
  AddSocialMediaMutationResult,
  SocialMediaType,
} from '#mutations/addSocialMediaMutation'
import {
  addUserDetailMutation,
  AddUserDetailInput,
  AddUserDetailMutationResult,
} from '#mutations/addUserDetailMutation'
import { removeSocialMediaMutation } from '#mutations/removeSocialMediaMutation'
import { removeUserDetailMutation } from '#mutations/removeUserDetailMutation'
import {
  updateUserMutation,
  UpdateUserInput,
  UpdateUserMutationResult,
} from '#mutations/updateUserMutation'
import { currentUserQuery } from '#queries/currentUserQuery'
import { searchUsersQuery } from '#queries/searchUsersQuery'
import { getInitials } from '#src/utils/getInitials'

export type { AddSocialMediaInput, SocialMediaType } from '#mutations/addSocialMediaMutation'
export type { AddUserDetailInput } from '#mutations/addUserDetailMutation'

export type UserInTable = {
  id: number
  role: 'VIEWER' | 'MODERATOR'
  name: string
  username: string
}

export type MyTable = {
  id: number
  name: string
  public: boolean
  users: UserInTable[]
}

export type UserDetailCategory = 'place' | 'work' | 'language' | 'education' | 'feeling'

export type UserDetail = {
  id: number
  category: UserDetailCategory
  text: string
}

export type UserAvailability = null | 'available' | 'partly_available' | 'busy'

export type SocialMedia = {
  id: number
  type: SocialMediaType
  link: string
}

export type UserWithProfile = {
  id: number
  name: string
  username: string
  avatar?: string
  table?: MyTable
  introduction?: string
  availability: UserAvailability
  details: UserDetail[]
  social: SocialMedia[]
}

export type CurrentUser = UserWithProfile & { referenceId: string }

export type SearchUser = {
  id: number
  name: string
  username: string
}

export type SearchUsersResult = {
  users: SearchUser[]
}

export const useUserStore = defineStore(
  'user',
  () => {
    const currentUser = ref<CurrentUser | null>(null)

    const { result: currentUserQueryResult, loading } = useQuery(
      currentUserQuery,
      {},
      {
        prefetch: false,
        fetchPolicy: 'no-cache',
      },
    )

    watch(currentUserQueryResult, (data: { currentUser: CurrentUser }) => {
      setCurrentUser(data.currentUser)
    })

    const getCurrentUser = computed(() => currentUser.value)

    const getMyTable = computed(() => currentUser.value?.table)

    const setMyTable = (table: MyTable) => {
      if (currentUser.value) {
        setCurrentUser({
          ...currentUser.value,
          table,
        })
      }
    }

    const getUsersInMyTable = computed(() => currentUser.value?.table?.users)

    const getCurrentUserAvatar = computed(() => currentUser.value?.avatar)

    const getCurrentUserInitials = computed(() => {
      const name = currentUser.value?.name
      if (name) return getInitials(name)
      return ''
    })

    const setCurrentUser = (user: CurrentUser) => {
      currentUser.value = user
    }

    const { mutate: updateUserMutationResult } = useMutation<UpdateUserMutationResult>(
      updateUserMutation,
      {
        fetchPolicy: 'no-cache',
      },
    )

    const updateUser = async (data: UpdateUserInput) => {
      const result = await updateUserMutationResult({ data })
      if (result?.data) {
        setCurrentUser(result.data.updateUser)
      }
    }

    const { mutate: addUserDetailMutationResult } = useMutation<AddUserDetailMutationResult>(
      addUserDetailMutation,
      {
        fetchPolicy: 'no-cache',
      },
    )

    const addUserDetail = async (userDetail: AddUserDetailInput) => {
      const result = await addUserDetailMutationResult({ data: userDetail })
      if (result?.data && currentUser.value?.id) {
        setCurrentUser({
          ...currentUser.value,
          details: [...currentUser.value.details, result.data.addUserDetail],
        })
      }
    }

    const { mutate: removeUserDetailMutationResult } = useMutation(removeUserDetailMutation, {
      fetchPolicy: 'no-cache',
    })

    const removeUserDetail = async (userDetailId: number) => {
      await removeUserDetailMutationResult({ id: userDetailId })
      if (currentUser.value?.id) {
        setCurrentUser({
          ...currentUser.value,
          details: currentUser.value.details.filter((d) => d.id !== userDetailId),
        })
      }
    }

    const { mutate: AddSocialMediaMutationResult } = useMutation<AddSocialMediaMutationResult>(
      addSocialMediaMutation,
      {
        fetchPolicy: 'no-cache',
      },
    )

    const addSocialMedia = async (socialMedia: AddSocialMediaInput) => {
      const result = await AddSocialMediaMutationResult({ data: socialMedia })
      if (result?.data && currentUser.value?.id) {
        setCurrentUser({
          ...currentUser.value,
          social: [...currentUser.value.social, result.data.addSocialMedia],
        })
      }
    }

    const { mutate: removeSocialMediaMutationResult } = useMutation(removeSocialMediaMutation, {
      fetchPolicy: 'no-cache',
    })

    const removeSocialMedia = async (socialMediaId: number) => {
      await removeSocialMediaMutationResult({ id: socialMediaId })
      if (currentUser.value?.id) {
        setCurrentUser({
          ...currentUser.value,
          social: currentUser.value.social.filter((s) => s.id !== socialMediaId),
        })
      }
    }

    const searchUsers = (name: string) => {
      const { result, loading, error } = useQuery(
        searchUsersQuery,
        { searchString: name },
        {
          fetchPolicy: 'no-cache',
        },
      )
      return { result, loading, error }
    }

    return {
      loading,
      currentUser,
      getCurrentUser,
      setCurrentUser,
      getCurrentUserInitials,
      getCurrentUserAvatar,
      getMyTable,
      setMyTable,
      getUsersInMyTable,
      searchUsers,
      updateUser,
      addUserDetail,
      addSocialMedia,
      removeUserDetail,
      removeSocialMedia,
    }
  },
  {
    persist: false,
  },
)

if (import.meta.hot) {
  import.meta.hot.accept(acceptHMRUpdate(useUserStore, import.meta.hot))
}
