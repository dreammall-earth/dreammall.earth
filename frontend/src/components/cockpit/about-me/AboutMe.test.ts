import { provideApolloClient } from '@vue/apollo-composable'
import { mount } from '@vue/test-utils'
import { createMockClient } from 'mock-apollo-client'
import { describe, expect, it, vi } from 'vitest'

import { currentUserQuery } from '#queries/currentUserQuery'
import { SocialMedia, UserDetail, useUserStore } from '#stores/userStore'

import AboutMe from './AboutMe.vue'

const currentUserQueryMock = vi.fn()

const mockClient = createMockClient()

mockClient.setRequestHandler(
  currentUserQuery,
  currentUserQueryMock.mockResolvedValue({
    data: {
      currentUser: {
        id: 666,
        name: 'Current User',
        username: 'currentUser',
        availability: 'available',
        introduction: 'Hello, I am the current user',
        details: [
          { id: 1, category: 'education', text: 'I am a student' },
          { id: 2, category: 'work', text: 'I am a worker' },
          { id: 3, category: 'work', text: 'I really like my job' },
          { id: 4, category: 'language', text: 'English' },
          { id: 5, category: 'place', text: 'Berlin' },
          { id: 6, category: 'feeling', text: 'Yeah!' },
        ] as UserDetail[],
        social: [
          {
            id: 1,
            type: 'instagram',
            link: 'https://instagram.com',
          },
          {
            id: 2,
            type: 'facebook',
            link: 'https://facebook.com',
          },
          {
            id: 3,
            type: 'linkedin',
            link: 'https://linkedin.com',
          },
          {
            id: 4,
            type: 'x',
            link: 'https://x.com',
          },
          {
            id: 5,
            type: 'xing',
            link: 'https://xing.com',
          },
          { id: 6, type: 'discord', link: 'https://discord.com' },
          { id: 7, type: 'telegram', link: 'https://t.me' },
          { id: 8, type: 'snapchat', link: 'https://snapchat.com' },
          { id: 9, type: 'reddit', link: 'https://reddit.com' },
          { id: 10, type: 'whatsapp', link: 'https://wa.me' },
          { id: 11, type: 'pintarest', link: 'https://pinterest.com' },
          { id: 12, type: 'linkedin', link: 'https://linkedin.com' },
        ] as SocialMedia[],
      },
    },
  }),
)

provideApolloClient(mockClient)

const setCurrentUser = (store: ReturnType<typeof useUserStore>) => {
  store.currentUser = {
    id: 666,
    referenceId: 'UQV6KSVD',
    name: 'Current User',
    username: 'currentUser',
    availability: 'available',
    introduction: 'Hello, I am the current user',
    details: [
      { id: 1, category: 'education', text: 'I am a student' },
      { id: 2, category: 'work', text: 'I am a worker' },
      { id: 3, category: 'work', text: 'I really like my job' },
      { id: 4, category: 'language', text: 'English' },
      { id: 5, category: 'place', text: 'Berlin' },
      { id: 6, category: 'feeling', text: 'Yeah!' },
    ] as UserDetail[],
    social: [
      {
        id: 1,
        type: 'instagram',
        link: 'https://instagram.com',
      },
      {
        id: 2,
        type: 'facebook',
        link: 'https://facebook.com',
      },
      {
        id: 3,
        type: 'linkedin',
        link: 'https://linkedin.com',
      },
      {
        id: 4,
        type: 'x',
        link: 'https://x.com',
      },
      {
        id: 5,
        type: 'xing',
        link: 'https://xing.com',
      },
      { id: 6, type: 'discord', link: 'https://discord.com' },
      { id: 7, type: 'telegram', link: 'https://t.me' },
      { id: 8, type: 'snapchat', link: 'https://snapchat.com' },
      { id: 9, type: 'reddit', link: 'https://reddit.com' },
      { id: 10, type: 'whatsapp', link: 'https://wa.me' },
      { id: 11, type: 'pintarest', link: 'https://pinterest.com' },
      { id: 12, type: 'linkedin', link: 'https://linkedin.com' },
    ] as SocialMedia[],
  }
}

describe('AboutMe', () => {
  const Wrapper = (props = {}) =>
    mount(AboutMe, {
      props,
    })

  it('renders', () => {
    const wrapper = Wrapper()
    const userStore = useUserStore()
    setCurrentUser(userStore)
    expect(wrapper.element).toMatchSnapshot()
  })

  /*
  it('updates User with new availability', () => {
    const wrapper = Wrapper()
    const userStore = useUserStore()
    setCurrentUser(userStore)
    const aboutMeView = wrapper.getComponent(AboutMeView)

    // eslint-disable-next-line @typescript-eslint/no-unsafe-call, @typescript-eslint/no-unsafe-member-access
    aboutMeView.vm.$emit('update-availability', 'busy')

    expect(userStore.updateUser).toHaveBeenCalledWith({
      name: 'Current User',
      availability: 'busy',
    })
  }) */
})
