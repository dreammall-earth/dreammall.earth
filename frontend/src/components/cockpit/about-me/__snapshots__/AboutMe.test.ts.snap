// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`AboutMe > renders 1`] = `
<div
  data-v-app=""
>
  
  
</div>
`;
